engine/src/main/resources/assets/shaders/chunk_frag.glsl

```diff
@@ -92,34 +92,41 @@ void main() {
     vec3 normalOpaque = normal;
     float shininess = 0.0;
 
-#if defined (NORMAL_MAPPING)
+#if defined (NORMAL_MAPPING) || defined (PARALLAX_MAPPING)
     // TODO: Calculates the tangent frame on the fly - this is absurdly costly... But storing
     // the tangent for each vertex in the chunk VBO might be not the best idea either.
-    vec3 dp1 = dFdx(vertexProjPos.xyz);
-    vec3 dp2 = dFdy(vertexProjPos.xyz);
-    vec2 duv1 = dFdx(gl_TexCoord[0].xy);
-    vec2 duv2 = dFdy(gl_TexCoord[0].xy);
-
-    vec3 dp2perp = cross(dp2, normal);
-    vec3 dp1perp = cross(normal, dp1);
-    vec3 tangent = dp2perp * duv1.x + dp1perp * duv2.x;
-    vec3 binormal = dp2perp * duv1.y + dp1perp * duv2.y;
-
-    float invMax = inversesqrt(max(dot(tangent,tangent), dot(binormal,binormal)));
-    mat3 tbn = mat3(tangent * invMax, binormal * invMax, normal);
+    // Screen-space coordinates happen to be a coordinate system to which it's possible to relate both view-space and texture coordinates, but are themselves irrelevant.
+    mat2x3 screenToView = mat2x3(dFdx(vertexViewPos.xyz), dFdy(vertexViewPos.xyz));
+    mat2   screenToUv   = mat2  (dFdx(gl_TexCoord[0].xy), dFdy(gl_TexCoord[0].xy)) / TEXTURE_OFFSET;
+    //mat2 uvToScreem = inverse(screenToUv);
+    //inverse is not available in GLSL 1.20, so calculate it manually:
```

Comment:
How about you wrap this into a small function at the beginning of the shader file? This way you don't have to resort to indentation to separate it from the rest of the code and we don't get distracted by it.